Refactor `DashToSeparator` Filter

ramchale opened the pull request:
|    Q          |   A
|-------------- | ------
| Documentation | no
| Bugfix        | no
| BC Break      | yes
| New Feature   | no
| RFC           | no
| QA            | no

### Description

Refactoring DashToSeparator based on #177

ramchale committed:
Remove inheritance chain from DashToSeparator

Signed-off-by: ramchale <legionsofbob@gmail.com>

ramchale commented:
Opening a Draft to seek feedback before I progress this any further. Could you let me know if this is broadly what you had in mind please?

ramchale commented on src/Word/DashToSeparator.php:
```diff
@@ -14,15 +17,42 @@
+        }
+
+        return $this->filterNormalizedValue(
+            array_map(static fn($item) => is_scalar($item) ? (string) $item : $item, $value)
```
While this causes a static analysis issue, I believe it is consistent with what is being done in AbstractFilter::applyFilterOnlyToStringableValuesAndStringableArrayValues . So changing it would likely introduce a change in behaviour

gsteel replied:
`AbstractFilter::applyFilterOnlyToStringableValuesAndStringableArrayValues` was extracted to [`ScalarOrArrayFilterCallback`](https://github.com/laminas/laminas-filter/blob/3.0.x/src/ScalarOrArrayFilterCallback.php)

In this case, I think you can just `return ScalarOrArrayFilterCallback::applyRecursively($value, fn (string $input): string => str_replace('-', $this->separator, $value));` and then drop `filterNormalizedValue`

ramchale commented on src/Word/DashToSeparator.php:
```diff
@@ -34,4 +64,16 @@ private function filterNormalizedValue($value)
     }
+
+    /** @return $this */
+    public function setSeparator(string $separator): self
```
These accessors aren't really needed for this, but removing them would potentially impact anyone currently using this class. Any preference on keeping them or not?

gsteel replied:
set|get Anything can be killed with 🔥

gsteel commented on src/Word/DashToSeparator.php:
```diff
@@ -14,15 +17,42 @@
  * }
  * @template TOptions of Options
- * @extends AbstractSeparator<TOptions>
+ * @implements FilterInterface<mixed>
```
```suggestion
 * @implements FilterInterface<string|array<array-key, string|mixed>>
```

gsteel commented on src/Word/DashToSeparator.php:
```diff
@@ -14,15 +17,42 @@
-final class DashToSeparator extends AbstractSeparator
+final class DashToSeparator implements FilterInterface
 {
+    protected string $separator = ' ';
```
```suggestion
    private readonly string $separator;
```

gsteel commented on src/Word/DashToSeparator.php:
```diff
@@ -14,15 +17,42 @@
 {
+    protected string $separator = ' ';
+
+    /**
+     * @param Options|string $separator Space by default
+     */
+    public function __construct(string|array $separator = ' ')
```
I'm going for consistent constructors throughout, so ideally:

```suggestion
    /** @param Options $separator */
    public function __construct(array $options = [])
```

The `Options` psalm type can also be sealed.

Inside the constructor, I would then apply the separator option to a private readonly property with `$this->separator = $options['separator'] ?? ' ';`

gsteel reviewed (changes requested):
Thanks @ramchale :)

Ideally, each filter refactor should be accompanied by changes to the docs too - Here's [an example](https://github.com/laminas/laminas-filter/pull/169/files) and a companion patch targeting the next 2.x minor with the relevant deprecations, like [this one](https://github.com/laminas/laminas-filter/pull/170/files)

ramchale committed:
Reduce code in DashToSeparator filter and add doc changes

- Use ScalarOrArrayFilterCallback::applyRecursively
- Remove setSeparator and getSeparator
- Add to migration doc
- Update usage doc to remove construction without Options array

Signed-off-by: ramchale <legionsofbob@gmail.com>

gsteel commented on docs/book/v3/word.md:
````diff
@@ -106,8 +106,7 @@ The following options are supported for `Laminas\Filter\Word\DashToSeparator`:
 ```php
-$filter = new Laminas\Filter\Word\DashToSeparator('+');
-// or new Laminas\Filter\Word\CamelCaseToSeparator(array('separator' => '+'));
+$filter = new Laminas\Filter\Word\DashToSeparator('separator' => '+');
````
```suggestion
$filter = new Laminas\Filter\Word\DashToSeparator(['separator' => '+']);
```

ramchale commented:
Thanks. I've updated that in line with the suggestions, let me know if further changes are needed. I'll open a corresponding deprecation branch against the current minor branch when I get a little bit of time

gsteel reviewed (changes requested):
Just one tweak to the docs, otherwise LGTM!

ramchale committed:
Update docs/book/v3/word.md

Co-authored-by: George Steel <george@netglue.uk>
Signed-off-by: Richard McHale <legionsofbob@gmail.com>

gsteel reviewed (approved):
Thank you @ramchale :)
